scope.go

```diff
@@ -182,52 +202,190 @@ func newRootScope(opts ScopeOptions, interval time.Duration) *scope {
 
 	s.hm.RLock()
-	for name, histogram := range s.histograms {
-		histogram.report(s.fullyQualifiedName(name), s.tags, r)
+	for name, histograms := range s.histograms {
+		for _, histogram := range histograms {
+			if histogram.report(s.fullyQualifiedName(name), s.tags, r, nowUnix) {
+				tracker.expiredHistograms[histogram] = struct{}{}
+			}
+		}
 	}
+
+	tracker.numHistograms = len(s.histograms)
 	s.hm.RUnlock()
 
+	s.clearExpiredMetrics(tracker)
+	if tracker.numMetrics() == 0 && globalNow().Unix() > s.createdUnix+s.getExpirySeconds() {
+		// All metrics for the scope have expired and it's been longer than the
+		// expiry period.
+		return true
+	}
+
 	r.Flush()
+	return false
 }
 
-func (s *scope) cachedReport(c CachedStatsReporter) {
+func (s *scope) cachedReport(c CachedStatsReporter) bool {
+	tracker := newScopeExpiryTracker()
+	nowUnix := globalNow().Unix()
+
 	s.cm.RLock()
-	for _, counter := range s.counters {
-		counter.cachedReport()
+	for _, counters := range s.counters {
+		for _, counter := range counters {
+			if counter.cachedReport(nowUnix) {
+				tracker.expiredCounters[counter] = struct{}{}
+			}
+		}
 	}
+
+	tracker.numCounters = len(s.counters)
 	s.cm.RUnlock()
 
 	s.gm.RLock()
-	for _, gauge := range s.gauges {
-		gauge.cachedReport()
+	for _, gauges := range s.gauges {
+		for _, gauge := range gauges {
+			if gauge.cachedReport(nowUnix) {
+				tracker.expiredGauges[gauge] = struct{}{}
+			}
+		}
 	}
+
+	tracker.numGauges = len(s.gauges)
 	s.gm.RUnlock()
 
-	// we do nothing for timers here because timers report directly to ths StatsReporter without buffering
+	// We do nothing for timers here because timers report directly to ths StatsReporter without buffering.
+	// If the scope is expired, new timers will be added to the fresh scope and refs to existing timers
+	// will continue to emit them directly to the reporters.
 
 	s.hm.RLock()
-	for _, histogram := range s.histograms {
-		histogram.cachedReport()
+	for _, histograms := range s.histograms {
+		for _, histogram := range histograms {
+			if histogram.cachedReport(nowUnix) {
+				tracker.expiredHistograms[histogram] = struct{}{}
+			}
+		}
 	}
+
+	tracker.numHistograms = len(s.histograms)
 	s.hm.RUnlock()
 
+	s.clearExpiredMetrics(tracker)
+	if tracker.numMetrics() == 0 && globalNow().Unix() > s.createdUnix+s.getExpirySeconds() {
+		// All metrics for the scope have expired and it's been longer than the
+		// expiry period.
+		return true
+	}
+
 	c.Flush()
+	return false
+}
+
+func (s *scope) clearExpiredMetrics(tracker *scopeExpiryTracker) {
+	var i int
+
+	if len(tracker.expiredCounters) > 0 {
+		s.cm.Lock()
+		for name, counters := range s.counters {
+			for _, counter := range counters {
+				if _, exists := tracker.expiredCounters[counter]; !exists ||
+					(exists && !counter.expired()) {
+					counters[i] = counter
+					i++
+				}
+			}
+
+			if i == 0 {
+				delete(s.counters, name)
+			} else {
+				s.counters[name] = counters[:i]
+			}
+
+			i = 0
```

Comment:
rather than resetting `i = 0`, which is easy to forget, sugget doing `var i int` inside of the for loop around `s.counters`.

same below for the other loops